### PR TITLE
[Enhancement] Use event driven to lazy instantiate drivers

### PR DESCRIPTION
**Pull request by ZiheLiu**
### Why I'm doing
As for the adaptive dop, the pipelines after ScanOperator / ExchangeSourceOperator will be instantiated in runtime, when the following conditions are satisfied:
1. The `CollectStatsSourceOperator` not blocked any more.
2. All the drivers of the dependent pipelines are finished.

There is a LazyInstantiateDriversOperator to check whether the above conditions are satisfied. Because this is a operator, it will run for a long time in the poller thread, especially when there are lots of LazyInstantiateDriversOperators in a query.

### What I'm doing

Use event driven instead of `LazyInstantiateDriversOperator`, since we could easily recognize when these two conditions are satisfied as follows:
1. **CsBlockingEvent** is finished when `CsSink` receives enough data or EOS.
2. **PipelineEvent** is finished when all the drivers of this pipeline are finished.
3. **CsSourceInitEvent** will be scheduled and instantiate pipelines, when all its dependent CsBlockingEvent and PipelineEvents are finished.

image

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/pipeline/adaptive/event.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+#include <atomic>
+#include <memory>
+#include <vector>
+
+#include "exec/pipeline/adaptive/adaptive_fwd.h"
+#include "exec/pipeline/pipeline_fwd.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks::pipeline {
```
Comment (satanson) [inline be/src/exec/pipeline/adaptive/event.h:25]: this feature is very excellent, however, can we re-design like this;

Event e0;
Event e1;

Event e2 = Event.dependsAnyOn(e0, e1);

Event e2 = Event.dependsAllOn(e0, e1);

e2 is ready only when e0 and e1 is finished.

File: be/src/exec/pipeline/adaptive/event.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/// Event is used to drive the execution of pipelines.
+/// Event::schedule is invoked when its dependencies are all finished.
+class Event : public std::enable_shared_from_this<Event> {
+public:
+    Event() = default;
+    virtual ~Event() = default;
+
+    /// This is not thread-safe.
+    virtual void schedule(RuntimeState* state) {}
```
Comment (kangkaisen) [inline be/src/exec/pipeline/adaptive/event.h:35]: What's schedule meaning of the event itself? maybe handle、process、dispatch?

File: be/src/exec/pipeline/adaptive/event.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    /// This is thread-safe.
+    void finish(RuntimeState* state);
+
+    /// This is thread-safe.
+    void finish_dependency(RuntimeState* state);
+
+    /// This is not thread-safe.
+    void add_dependency(Event* event);
```
Comment (kangkaisen) [inline be/src/exec/pipeline/adaptive/event.h:44]: Some interfaces are thread-safe, some not. which is confusing. how do we make true thread safe for the non-thread safe interfaces.

File: be/src/exec/pipeline/adaptive/event.h (hunk 2 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/adaptive/event.h:35]: `schedule` means this event will schedule some pipeline drivers to execute. `process` is better.

File: be/src/exec/pipeline/adaptive/event.h (hunk 3 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/adaptive/event.h:44]: non-thread safe interfaces contain `add_dependency` and `schedule`.
- `add_dependency` is only called when the pipeline graph is being built, so it guarantee execution in the single thread.
- `schedule` is call once only when it is ready.

**Review by satanson — APPROVED** [review]

File: be/src/exec/pipeline/fragment_context.h [hunk 4]
```diff
@@ -45,6 +45,13 @@ namespace pipeline {
 using RuntimeFilterPort = starrocks::RuntimeFilterPort;
 using PerDriverScanRangesMap = std::map<int32_t, std::vector<TScanRangeParams>>;
 
+// clang-format off
```
Comment (liuyehcf) [inline be/src/exec/pipeline/fragment_context.h:48]: this comment can turn clang-format off? it seems really normal.

**Review by liuyehcf — APPROVED** [review]

File: be/src/exec/pipeline/fragment_context.h (hunk 4 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/fragment_context.h:48]: Yes. clang-format has bug for multiple lines of `concept`, so disable it.

**Pull request merged**